src/DefaultBuilder/src/Configuration.cs

```diff
@@ -19,69 +19,105 @@ namespace Microsoft.AspNetCore.Builder
     public sealed class Configuration : IConfigurationRoot, IConfigurationBuilder, IDisposable
     {
         private readonly ConfigurationSources _sources;
-        private ConfigurationRoot _configurationRoot;
+        private readonly ConfigurationBuilderProperties _properties;
 
+        private readonly object _providerLock = new();
+        private readonly List<IConfigurationProvider> _providers = new();
+        private readonly List<IDisposable> _changeTokenRegistrations = new();
         private ConfigurationReloadToken _changeToken = new();
-        private IDisposable? _changeTokenRegistration;
 
         /// <summary>
         /// Creates an empty mutable configuration object that is both an <see cref="IConfigurationBuilder"/> and an <see cref="IConfigurationRoot"/>.
         /// </summary>
         public Configuration()
         {
             _sources = new ConfigurationSources(this);
+            _properties = new ConfigurationBuilderProperties(this);
 
             // Make sure there's some default storage since there are no default providers.
             this.AddInMemoryCollection();
 
-            Update();
+            NotifySourceAdded(_sources[0]);
         }
 
-        /// <summary>
-        /// Automatically update the <see cref="IConfiguration"/> on <see cref="IConfigurationBuilder"/> changes.
-        /// If <see langword="false"/>, <see cref="Update()"/> will manually update the <see cref="IConfiguration"/>.
-        /// </summary>
-        internal bool AutoUpdate { get; set; } = true;
+        /// <inheritdoc/>
+        public string? this[string key]
+        {
+            get
+            {
+                lock (_providerLock)
+                {
+                    for (int i = _providers.Count - 1; i >= 0; i--)
```

Comment:
The implementation of the indexer should be identical to ConfigurationRoot right?  If its too complicated to have a good way to enforce (maybe a new ConfigurationProviderList in extensions?), perhaps put a comment pointing to ConfigurationRoot saying this should stay in sync for posterity?